proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+  /// during the execution of `withUnsafeBytes(_:)`.
+  /// Do not store or return the pointer for later use.
+  ///
+  /// - Parameter body: A closure with an `UnsafeRawBufferPointer`
+  ///   parameter that points to the viewed contiguous storage.
+  ///   If `body` has a return value, that value is also
+  ///   used as the return value for the `withUnsafeBytes(_:)` method.
+  ///   The closure's parameter is valid only for the duration of
+  ///   its execution.
+  /// - Returns: The return value of the `body` closure parameter.
+  func withUnsafeBytes<Result>(
+    _ body: (_ buffer: UnsafeRawBufferPointer) -> Result
+  ) -> Result
+}
+```
+
+### RawSpan
+
+In addition to `Span<T>`, we propose the addition of `RawSpan`, to represent heterogenously-typed values in contiguous memory. `RawSpan` is similar to `Span<T>`, but represents _untyped_ initialized bytes. `RawSpan` is a specialized type that intends to support parsing and decoding applications, as well as applications where heavily-used code paths require concrete types as much as possible. Its API supports extracting sub-spans, along with the operations `load(as:)` and `loadUnaligned(as:)`.
+
+#### Complete `RawSpan` API:
+
+```swift
+public struct RawSpan: Copyable, ~Escapable {
+  internal var _start: RawSpan.Index
+  internal var _count: Int
+}
+```
+
+##### Initializing a `RawSpan`:
+
+```swift
+extension RawSpan {
+  /// Unsafely create a `RawSpan` over initialized memory.
+  ///
+  /// The memory in `buffer` must be owned by the instance `owner`,
+  /// meaning that as long as `owner` is alive the memory will remain valid.
+  ///
+  /// - Parameters:
+  ///   - buffer: an `UnsafeRawBufferPointer` to initialized memory.
+  ///   - owner: a binding whose lifetime must exceed that of
+  ///            the newly created `RawSpan`.
+  public init?<Owner: ~Copyable & ~Escapable>(
+    unsafeBytes buffer: UnsafeBufferPointer<Element>,
+    owner: borrowing Owner
+  ) -> dependsOn(owner) Self?
+
+  /// Unsafely create a `RawSpan` over initialized memory.
+  ///
+  /// The memory over `count` bytes starting at
+  /// `pointer` must be owned by the instance `owner`,
+  /// meaning that as long as `owner` is alive the memory will remain valid.
+  ///
+  /// - Parameters:
+  ///   - pointer: a pointer to the first initialized element.
+  ///   - count: the number of initialized elements in the view.
+  ///   - owner: a binding whose lifetime must exceed that of
+  ///            the newly created `RawSpan`.
+  public init<Owner: ~Copyable & ~Escapable>(
+    unsafeRawPointer pointer: UnsafeRawPointer,
+    count: Int,
+    owner: borrowing Owner
+  ) -> dependsOn(owner) Self
+
+  /// Create a `RawSpan` over the memory represented by a `Span<T>`
+  ///
+  /// - Parameters:
+  ///   - span: An existing `Span<T>`, which will define both this
+  ///           `RawSpan`'s lifetime and the memory it represents.
+  @inlinable @inline(__always)
```

Comment:
Nit: `@inline(__always)` is an implementation detail; it should not appear as part of the proposed API.

Reply:
Yeah that's a copy-pasting error
```suggestion
```